### PR TITLE
Connection established and closed callbacks

### PR DESCRIPTION
**Pull request by tewinget**
builds on #39

### CODE REVIEW

File: include/quic/error.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        CONN_DRAIN_NO_ERR = (1ULL << 62) - 94,
+
+        DATAGRAM_EXCEPTION = (1ULL << 62) - 64,
+
+        STREAM_EXCEPTION = (1ULL << 62) - 32,
+        STREAM_CONNECTION_EXPIRED = (1ULL << 62) - 31,
+
+        CONN_WRITE_CLOSE_FAIL = (1ULL << 62),
+        CONN_SEND_CLOSE_FAIL = (1ULL << 62) + 1
+    };
```
Comment (jagerman) [inline include/quic/error.hpp:19]: This needs a comment as to why 2^62 ± N is being used here as the underlying value.

Comment (dr7ana) [reply include/quic/error.hpp:19]: Resolved by off-band discussion, to-be fixed by Tom's next commits

File: tests/009-close.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        Network test_net{};
+
+        auto server_tls = GNUTLSCreds::make("./serverkey.pem"s, "./servercert.pem"s, "./clientcert.pem"s);
+        auto client_tls = GNUTLSCreds::make("./clientkey.pem"s, "./clientcert.pem"s, "./servercert.pem"s);
+
+        opt::local_addr server_local{};
+        opt::local_addr client_local{};
+
+        SECTION("Connection Closes with correct error code")
```
Comment (dr7ana) [inline tests/009-close.cpp:30]: looks good to me! one question: are you planning on adding a second section to this test case?

Comment (tewinget) [reply tests/009-close.cpp:30]: nah, just couldn't remember if it had to have a section or not and figured it wouldn't hurt to put it regardless

Comment (dr7ana) [reply tests/009-close.cpp:30]: agreed

**Review by dr7ana — APPROVED** [review]

File: src/connection.cpp [hunk 3]
```diff
@@ -234,21 +264,20 @@ namespace oxen::quic
         event_active(packet_io_trigger.get(), 0, 0);
     }
 
-    void Connection::close_connection()
+    void Connection::close_connection(uint64_t error_code)
     {
-        _endpoint.call([this]() { _endpoint.close_connection(*this); });
+        _endpoint.call([this, error_code]() { _endpoint.close_connection(*this, io_error{error_code}); });
```
Comment (jagerman) [inline src/connection.cpp:269]: Perhaps:
```suggestion
        _endpoint.call([this, err=io_error{error_code}]() { _endpoint.close_connection(*this, err); });
```

File: src/connection.cpp [hunk 4]
```diff
@@ -757,19 +776,24 @@ namespace oxen::quic
             return;
         }
 
-        auto delta = exp_ns * 1ns - ts.time_since_epoch();
+        auto delta = std::chrono::nanoseconds(exp_ns * 1ns) - ts.time_since_epoch();
```
Comment (jagerman) [inline src/connection.cpp]:
```suggestion
        auto delta = std::chrono::nanoseconds{exp_ns} - ts.time_since_epoch();
```

Comment (jagerman) [reply src/connection.cpp]: Or:
```suggestion
        auto delta = static_cast<int64_t>(exp_ns) * 1ns - ts.time_since_epoch();
```

File: src/endpoint.cpp [hunk 5]
```diff
@@ -105,10 +117,24 @@ namespace oxen::quic
...
-        conn.call_close_cb();
         conn.halt_events();
-
         conn.set_draining();
+
+        const auto* err = ngtcp2_conn_get_ccerr(conn);
+
+        std::string_view reason = "None"sv;
+        if (err->reasonlen > 0)
+            reason = std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen};
```
Comment (jagerman) [inline src/endpoint.cpp]:
```suggestion
        auto reason = err->reasonlen > 0
            ? std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen}
            : "None"sv;
```

File: src/endpoint.cpp [hunk 6]
```diff
@@ -105,10 +117,24 @@ namespace oxen::quic
...
+
+        log::debug(log_cat, "Draining connection (CID: {}), Reason: {}", conn.scid(), reason);
```
Comment (jagerman) [inline src/endpoint.cpp]:
```suggestion
        log::debug(log_cat, "Draining connection (CID: {}), Reason: {}", conn.scid(),
            err->reason ? std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen} : "None"sv);
```

File: src/endpoint.cpp [hunk 7]
```diff
@@ -154,39 +180,58 @@ namespace oxen::quic
...
-    void Endpoint::close_connection(Connection& conn, int code, std::string_view msg)
+    void Endpoint::drop_connection(Connection& conn)
+    {
+        const auto* err = ngtcp2_conn_get_ccerr(conn);
+
+        std::string_view reason = "None"sv;
+        if (err->reasonlen > 0)
+            reason = std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen};
+
+        log::debug(log_cat, "Dropping connection (CID: {}), Reason: {}", conn.scid(), reason);
```
Comment (jagerman) [inline src/endpoint.cpp]:
```suggestion
        log::debug(log_cat, "Dropping connection (CID: {}), Reason: {}", conn.scid(),
            err->reason ? std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen} : "None"sv);
```

File: src/endpoint.cpp [hunk 8]
```diff
@@ -198,8 +243,20 @@ namespace oxen::quic
...
+            log::trace(log_cat, "{} Calling Connection closed callback", conn.is_inbound() ? "server" : "client");
+            connection_close_cb(conn, ec.code());
+        }
+
         ngtcp2_ccerr err;
-        ngtcp2_ccerr_set_liberr(&err, code, reinterpret_cast<uint8_t*>(const_cast<char*>(msg.data())), msg.size());
+        ngtcp2_ccerr_default(&err);
+        if (ec.is_ngtcp2)
+            ngtcp2_ccerr_set_liberr(
+                    &err, ec.ngtcp2_code(), reinterpret_cast<uint8_t*>(const_cast<char*>(msg.data())), msg.size());
```
Comment (jagerman) [inline src/endpoint.cpp]:
```suggestion
                    &err, ec.ngtcp2_code(), reinterpret_cast<const uint8_t*>(msg.data()), msg.size());
```

File: src/endpoint.cpp [hunk 9]
```diff
@@ -198,8 +243,20 @@ namespace oxen::quic
...
+        else
+            ngtcp2_ccerr_set_application_error(
+                    &err, ec.code(), reinterpret_cast<uint8_t*>(const_cast<char*>(msg.data())), msg.size());
```
Comment (jagerman) [inline src/endpoint.cpp]:
```suggestion
                    &err, ec.code(), reinterpret_cast<const uint8_t*>(msg.data()), msg.size());
```

**Review by jagerman — APPROVED** [review]

**Pull request merged**